File: BaseLibrary/communication/protocol/ProtocolFactory.cs
using System.Collections.Generic;

namespace BaseLibrary.communication.protocol {
    /// <summary>
    /// Factory for getting implementation of protocol by its name.
    /// </summary>
    public class ProtocolFactory {

        private readonly IDictionary<string, AProtocol> protocolsMap = new Dictionary<string, AProtocol>();

        /// <summary>
        /// Return some protocol with name from protocolLabels.
        /// </summary>
        /// <param name="protocolLabels"></param>
        /// <returns></returns>
        public AProtocol GetProtocol(params string[] protocolLabels) {
            string trash;
            return GetProtocol(out trash, protocolLabels);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="selectedProtocolLabel"></param>
        /// <param name="protocolLabels"></param>
        /// <returns></returns>
        public AProtocol GetProtocol(out string selectedProtocolLabel, params string[] protocolLabels) {
            AProtocol protocol;
            foreach (string protocolLabel in protocolLabels) {
                if (protocolsMap.TryGetValue(protocolLabel, out protocol)) {
                    selectedProtocolLabel = protocolLabel;
                    return protocol;
                }
            }
            selectedProtocolLabel = null;
            return null;
        }

        /// <summary>
        /// Register protocol by its name.
        /// </summary>
        /// <param name="protocolLabel"></param>
        /// <param name="protocol"></param>
        public void RegisterProtocol(string protocolLabel, AProtocol protocol) {
            protocolsMap.Add(protocolLabel, protocol);
        }
    }
}

File: Java-Client/RobotBattlefield-Client-Java/src/BaseLibrary/communication/protocol/ProtocolFactory.java
package BaseLibrary.communication.protocol;

import java.util.HashMap;

import BaseLibrary.utils.Holder;

public class ProtocolFactory {

	private final HashMap<String, AProtocol> protocolsMap = new HashMap<>();

	public AProtocol GetProtocol(String... protocolLabels) {
		Holder<String> trash = new Holder<>();

		return GetProtocol(trash, protocolLabels);
	}

	public AProtocol GetProtocol(Holder<String> selectedProtocolLabel, String... protocolLabels) {
		AProtocol protocol;
		for (String protocolLabel : protocolLabels) {
			if ((protocol = protocolsMap.get(protocolLabel)) != null) {
				selectedProtocolLabel.value = protocolLabel;
				return protocol;
			}
		}
		return null;
	}

	public void RegisterProtocol(String protocolLabel, AProtocol protocol) {
		protocolsMap.put(protocolLabel, protocol);
	}
	
}
